### PR TITLE
Feature/repo files

### PR DESCRIPTION
**Pull request by zlusch95**
Pre hooks for diff module including pull requests, pull request files and repository files

### CODE REVIEW

**Comment by DieserFelix** [comment]
Looks good to me

File: src/entities/diffs/diff.module.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          schema.pre<Diff>('validate', async function (this: Diff) {
+            this.pullRequest = (
+              await pullRequestService.readOrCreate(this.pullRequest)
+            )._id;
+            if (this.pullRequestFiles) {
+              for (let i = 0; i < this.pullRequestFiles.length; i++) {
+                this.pullRequestFiles[i] = await (
+                  await pullRequestFileService.readOrCreate(
+                    this.pullRequestFiles[i],
+                  )
```
Comment (janniclas) [inline src/entities/diffs/diff.module.ts:41]: why do we need this double await here? 
the code looks more or less identical to the creation of repositoryFiles in line 48 and there a single await is sufficient

File: src/entities/diffs/lib/pullRequestFileQuerier.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  repoIdent: RepositoryIdentifier,
+  pullRequest: PullRequest,
+) {
+  const pullRequestFiles: PullRequestFile[] = [];
+  let page: PullRequestFile[] = [];
+  let pageNumber = 1;
+
+  do {
+    page = await queryPullRequestFilePage(repoIdent, pullRequest, pageNumber);
+    pullRequestFiles.push(...page);
```
Comment (janniclas) [inline src/entities/diffs/lib/pullRequestFileQuerier.ts:16]: I believe we need to use the generator pattern here as we do in the pullRequestQuerier. Else this array will easily overflow

**Review by janniclas — CHANGES_REQUESTED** [review]
I believe the only really necessary change is the one regarding the generator

File: src/entities/diffs/diff.module.ts (hunk 1 above)
Comment (DieserFelix) [reply src/entities/diffs/diff.module.ts:41]: It is sufficient here, too.

File: src/entities/diffs/lib/pullRequestFileQuerier.ts (hunk 2 above)
Comment (DieserFelix) [reply src/entities/diffs/lib/pullRequestFileQuerier.ts:16]: In general, I tend to agree. However, we need the full array of changed files in order to calculate their subsets.  
I suppose we could go for a generator here and accumulate the files in the CouplingOfComponents module. That way, we avoid imposing the memory overhead to every KPI using PR files.

**Pull request merged**